amalgamate.mjs

```diff
@@ -0,0 +1,264 @@
+    } catch {
+        return [];
+    }
+    return buf
+        .toString()
+        .split("\n")
+        .filter(Boolean)
+        ;
+}
+
+const filesWithErrorsPath = amalgamatedPath("files_with_errors.txt");
+const filesWithErrors = new Set((await readLines(filesWithErrorsPath))
+    .map(path => pathLib.relative(".", path))
+);
+console.warn(`skipping ${filesWithErrors.size} files with errors`);
+await fsp.writeFile(filesWithErrorsPath, [...filesWithErrors].join("\n"));
+
+const flagsClangDoesntKnow = new Set([
+    // "-Wno-format-truncation",
+    // "-Wno-stringop-truncation",
+]);
+
+const skippedFiles = new Set([
+    // We use the real definitions in "psp/fsw/pc-linux/src/cfe_psp_start.c".
+    "osal/src/bsp/shared/src/bsp_default_app_startup.c",
+    "osal/src/bsp/shared/src/bsp_default_app_run.c",
+]);
+
+async function amalgamate({ redefinitions }) {
+    // console.log(redefinitions);
+
+    const commands = JSON.parse((await fsp.readFile("compile_commands.json")).toString())
+        .map(({ directory, file, output, command, arguments: args }) => {
+            const relativePath = pathLib.relative(".", file);
+
+            args = args ?? command.split(/ +/g);
+            command = args.join(" ");
+
+            const flags = args
+                .slice(
+                    ["cc"].length, // Slice off the compiler invocation (not a flag).
+                    -["-o", "obj", "-c", "src"].length, // Slice off the file-specific args that aren't flags.
+                )
+                .filter(flag => !flagsClangDoesntKnow.has(flag))
+                ;
+
+            const definePrefix = "-D";
+            const defineSep = "=";
+            let commonDefines = new Set();
+            const setCommonDefines = (defines) => {
+                commonDefines = defines;
+            };
+            const existingDefines = () => flags
+                .filter(flag => flag.startsWith(definePrefix))
+                .filter(define => !commonDefines.has(define))
+                .map(flag => {
+                    const [name, ...rest] = flag
+                        .slice(definePrefix.length)
+                        .split(defineSep);
+                    let value = rest.join(defineSep);
+                    value = value.replaceAll('\\"', '\"');
+                    return { name, value, flag };
+                });
+            const pathIdentifier = relativePath.replace(/\W+/g, '_');
+            const defines = () => [
+                ...existingDefines(),
+                ...(redefinitions[file] ?? [])
+                    // ...[...new Set(Object.values(redefinitions).flat())]
+                    .map(name => ({ name, value: `${pathIdentifier}__${name}` })),
+            ];
+
+            const flagsWithoutDefines = flags.filter(flag => !flag.startsWith(definePrefix));
+
+            const includeLines = () => [
+                ...defines().map(({ name, value }) => `#define ${name} ${value}`),
```

Comment:
The macro might have already been defined elsewhere. I think we should do something like (see #2)
```
#pragma push_macro("${name}")
#undef ${name}
#define ${name} ${value}
...
#undef ${name}
#pragma pop_macro("${name}")
```

Reply:
How does `{push,pop}_macro` work?  If it works the way I assume it does, then yes, this seems more robust and handles macros, too.